lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
   }
 };
 
+struct PoolingToPool2dPattern : public OpConversionPattern<ttir::PoolingOp> {
+public:
+  using OpConversionPattern<ttir::PoolingOp>::OpConversionPattern;
+
+  std::vector<int64_t> getSpatialDims(ttir::PoolingOp op) const {
+    std::vector<int64_t> spatial_dims;
+    for (int64_t i = 0;
+         i < static_cast<int64_t>(op.getWindowDimensions().size()); i++) {
+      if (op.getWindowDimensions()[i] > 1) {
+        spatial_dims.push_back(i);
+      }
+    }
+    return spatial_dims;
+  }
+
+  LogicalResult canDecompose2DPoolingOp(ttir::PoolingOp op) const {
+
+    // Window dimensions must be 4 in length
+    if (op.getWindowDimensions().size() != 4) {
+      return failure();
+    }
+
+    // Window strides must be 4 in length
+    if (op.getWindowStrides().size() != 4) {
+      return failure();
+    }
+
+    // Operand rank(s) must be 4
+    for (Value operand : op.getInputs()) {
+      auto operand_type = mlir::cast<mlir::RankedTensorType>(operand.getType());
+      if (operand_type.getRank() != 4) {
+        return failure();
+      }
+    }
+
+    // Exactly two of the window dimensions must be greater than 1
+    std::vector<int64_t> true_window_dimensions_indices = getSpatialDims(op);
+
+    if (true_window_dimensions_indices.size() != 2) {
+      return failure();
+    }
+
+    // Exactly two of the window strides must be greater than 1
+    std::vector<int64_t> true_window_stride_indices;
+    for (int64_t i = 0; i < static_cast<int64_t>(op.getWindowStrides().size());
+         i++) {
+      if (op.getWindowStrides()[i] > 1) {
+        true_window_stride_indices.push_back(i);
+      }
+    }
+
+    if (true_window_stride_indices.size() != 2) {
+      return failure();
+    }
+
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // Padding must be 8 in length
+    if (op.getPadding().size() != 8) {
+      return failure();
+    }
+
+    return success();
+  }
+
+  template <typename PoolOpType>
+  void rewritePool2d(ttir::PoolingOp op, OpAdaptor adaptor,
+                     ConversionPatternRewriter &rewriter) const {
+
+    const int64_t SPATIAL_H = -3;
+    const int64_t SPATIAL_W = -2;
+    const int64_t NON_SPATIAL = -1;
+
+    auto inputType =
+        mlir::cast<RankedTensorType>(adaptor.getInputs()[0].getType());
+    assert(inputType.getRank() == 4 && "Input must be 4D tensor");
+    std::vector<int64_t> desired_layout(inputType.getRank(), NON_SPATIAL);
+    desired_layout[inputType.getRank() - 3] = SPATIAL_H;
+    desired_layout[inputType.getRank() - 2] = SPATIAL_W;
```

Comment:
If rank is 4, that'd put H/W dims into second and third place, like this [x, H, W, x]? Oh, it's NHWC?